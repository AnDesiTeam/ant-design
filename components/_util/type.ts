// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#predefined-conditional-types
export type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
