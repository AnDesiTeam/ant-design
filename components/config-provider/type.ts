import { ValidateMessages } from 'rc-field-form/lib/interface';
import { RenderEmptyHandler } from './renderEmpty';
import { CSPConfig, DirectionType } from './context';
import { RequiredMark } from '../form/Form';
import { Locale } from '../locale-provider';
import { SizeType } from './SizeContext';

export interface ConfigProviderProps {
  getTargetContainer?: () => HTMLElement;
  getPopupContainer?: (triggerNode?: HTMLElement) => HTMLElement;
  prefixCls?: string;
  iconPrefixCls?: string;
  children?: React.ReactNode;
  renderEmpty?: RenderEmptyHandler;
  csp?: CSPConfig;
  autoInsertSpaceInButton?: boolean;
  form?: {
    validateMessages?: ValidateMessages;
    requiredMark?: RequiredMark;
  };
  input?: {
    autoComplete?: string;
  };
  locale?: Locale;
  pageHeader?: {
    ghost: boolean;
  };
  componentSize?: SizeType;
  direction?: DirectionType;
  space?: {
    size?: SizeType | number;
  };
  virtual?: boolean;
  dropdownMatchSelectWidth?: boolean;
}
