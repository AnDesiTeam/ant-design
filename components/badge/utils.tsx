import { PresetColorTypes } from '../_util/colors';
import type { PresetColorType } from '../_util/colors';

// eslint-disable-next-line import/prefer-default-export
export function isPresetColor(color?: string): boolean {
  return PresetColorTypes.includes(color as PresetColorType);
}
