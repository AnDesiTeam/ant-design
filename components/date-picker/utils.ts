import * as moment from 'moment';

export function formatDate(
  value: moment.Moment | undefined | null,
  format: string | string[] | ((moment: moment.Moment | undefined) => string),
): string {
  if (!value) {
    return '';
  }
  if (Array.isArray(format)) {
    format = format[0];
  }

  if (typeof format === 'function') {
    format = format(value);
  }

  return value.format(format);
}
