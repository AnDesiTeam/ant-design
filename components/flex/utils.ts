import classNames from 'classnames';

import type { FlexProps } from './interface';

export const flexDirectionValues = ['row', 'row-reverse', 'column', 'column-reverse'] as const;

export const flexWrapValues = ['wrap', 'nowrap', 'wrap-reverse'] as const;

export const justifyContentValues = [
  'flex-start',
  'flex-end',
  'start',
  'end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
  'stretch',
  'normal',
  'left',
  'right',
] as const;

export const alignItemsValues = [
  'center',
  'start',
  'end',
  'flex-start',
  'flex-end',
  'self-start',
  'self-end',
  'baseline',
  'normal',
  'stretch',
] as const;

const genClsWrap = (prefixCls: string, wrap: FlexProps['wrap']) => {
  const wrapCls: Record<PropertyKey, boolean> = {};
  flexWrapValues.forEach((cssKey) => {
    wrapCls[`${prefixCls}-wrap-${cssKey}`] = wrap === cssKey;
  });
  return wrapCls;
};

const genClsAlign = (prefixCls: string, align: FlexProps['align']) => {
  const alignCls: Record<PropertyKey, boolean> = {};
  alignItemsValues.forEach((cssKey) => {
    alignCls[`${prefixCls}-align-${cssKey}`] = align === cssKey;
  });
  return alignCls;
};

const genClsJustify = (prefixCls: string, justify: FlexProps['justify']) => {
  const justifyCls: Record<PropertyKey, boolean> = {};
  justifyContentValues.forEach((cssKey) => {
    justifyCls[`${prefixCls}-justify-${cssKey}`] = justify === cssKey;
  });
  return justifyCls;
};

const genClsDirection = (prefixCls: string, direction: FlexProps['direction']) => {
  const directionCls: Record<PropertyKey, boolean> = {};
  flexDirectionValues.forEach((cssKey) => {
    directionCls[`${prefixCls}-direction-${cssKey}`] = direction === cssKey;
  });
  return directionCls;
};

const genMergedAlignCls = (
  prefixCls: string,
  align: FlexProps['align'],
  direction: FlexProps['direction'],
) => {
  if (align) {
    return genClsAlign(prefixCls, align);
  }
  return {
    ...genClsAlign(prefixCls, align),
    [`${prefixCls}-align-stretch`]: direction === 'column' || direction === 'column-reverse',
  };
};

const createFlexClassNames = (
  prefixCls: string,
  props: Pick<FlexProps, 'direction' | 'wrap' | 'justify' | 'align'>,
) => {
  const { direction, wrap, justify, align } = props;
  return classNames({
    ...genClsWrap(prefixCls, wrap),
    ...genClsJustify(prefixCls, justify),
    ...genClsDirection(prefixCls, direction),
    ...genMergedAlignCls(prefixCls, align, direction),
  });
};

export default createFlexClassNames;
