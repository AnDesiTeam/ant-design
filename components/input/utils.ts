import type { ReactNode } from 'react';
import type { InputProps } from './Input';

// eslint-disable-next-line import/prefer-default-export
export function hasPrefixSuffix(props: {
  prefix?: ReactNode;
  suffix?: ReactNode;
  allowClear?: InputProps['allowClear'];
}) {
  return !!(props.prefix || props.suffix || props.allowClear);
}
