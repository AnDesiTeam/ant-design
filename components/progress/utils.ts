import { presetPrimaryColors } from '@ant-design/colors';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './progress';
import warning from '../_util/warning';

export function validProgress(progress?: number) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    warning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export const getPercentage = ({ percent, success, successPercent }: ProgressProps) => {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
};

export const getStrokeColor = ({ success = {}, strokeColor }: Partial<CircleProps>): (
  | string
  | Record<PropertyKey, string>
)[] => {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
};

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
  },
): [number, number] => {
  let width: number = -1;
  let height: number = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = size;
    }
    width *= steps;
  } else if (type === 'line') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = size === 'small' ? 6 : 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = size;
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      width = size[0] ?? size[1] ?? 120;
      height = size[0] ?? size[1] ?? 120;
    }
  }
  return [width, height];
};
