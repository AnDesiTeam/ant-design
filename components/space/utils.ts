import type { SpaceSize } from '.';
import type { SizeType } from '../config-provider/SizeContext';

export const isPresetSize = (size: SpaceSize): size is SizeType => {
  if (!size) {
    return false;
  }
  return typeof size === 'string' && ['small', 'middle', 'large'].includes(size);
};

export const isValidNumber = (size: SpaceSize): size is number => {
  if (!size) {
    // 这里故意排除 size = 0 的情况，因为 gap 属性在 CSS 中的默认值就是 0，所以如果用户传0进来的话，我们可以直接忽略
    return false;
  }
  return typeof size === 'number' && !Number.isNaN(size);
};
