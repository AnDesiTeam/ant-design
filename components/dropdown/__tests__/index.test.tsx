import React from 'react';
import type { TriggerProps } from 'rc-trigger';
import Dropdown from '..';
import Menu from '../../menu';
import type { DropDownProps } from '..';
import mountTest from '../../../tests/shared/mountTest';
import rtlTest from '../../../tests/shared/rtlTest';
import { act, fireEvent, render, waitFakeTimer } from '../../../tests/utils';

let triggerProps: TriggerProps;

jest.mock('rc-trigger', () => {
  let Trigger = jest.requireActual('rc-trigger/lib/mock');
  Trigger = Trigger.default || Trigger;
  const h: typeof React = jest.requireActual('react');

  return {
    default: h.forwardRef<unknown, TriggerProps>((props, ref) => {
      triggerProps = props;
      return h.createElement(Trigger, { ref, ...props });
    }),
    __esModule: true,
  };
});

describe('Dropdown', () => {
  const items = [
    {
      label: 'foo',
      key: '1',
    },
  ];

  mountTest(() => (
    <Dropdown menu={{ items }}>
      <span />
    </Dropdown>
  ));

  rtlTest(() => (
    <Dropdown menu={{ items }}>
      <span />
    </Dropdown>
  ));

  it('overlay is function and has custom transitionName', () => {
    const { asFragment } = render(
      <Dropdown overlay={() => <div>menu</div>} transitionName="move-up" open>
        <button type="button">button</button>
      </Dropdown>,
    );
    expect(Array.from(asFragment().childNodes)).toMatchSnapshot();
  });

  it('overlay is string', () => {
    const { asFragment } = render(
      <Dropdown overlay={'string' as any} open>
        <button type="button">button</button>
      </Dropdown>,
    );
    expect(Array.from(asFragment().childNodes)).toMatchSnapshot();
  });

  it('should render custom dropdown correctly', () => {
    const { asFragment } = render(
      <Dropdown
        open
        menu={{ items }}
        dropdownRender={menu => (
          <div>
            {menu}
            <div className="dropdown-custom-node">CUSTOM NODE</div>
          </div>
        )}
      >
        <button type="button">button</button>
      </Dropdown>,
    );
    expect(Array.from(asFragment().childNodes)).toMatchSnapshot();
  });

  it('support Menu expandIcon', async () => {
    jest.useFakeTimers();
    const props: DropDownProps = {
      menu: {
        items: [
          {
            label: 'foo',
            key: '1',
          },
          {
            label: 'SubMenu',
            key: 'submenu',
            children: [
              {
                label: 'foo',
                key: '1',
              },
            ],
          },
        ],
        expandIcon: <span id="customExpandIcon" />,
      },
      open: true,
      getPopupContainer: node => node,
    };

    const { container } = render(
      <Dropdown {...props}>
        <button type="button">button</button>
      </Dropdown>,
    );
    await waitFakeTimer();
    expect(container.querySelectorAll('#customExpandIcon').length).toBe(1);
    jest.useRealTimers();
  });

  it('should warn if use topCenter or bottomCenter', () => {
    const error = jest.spyOn(console, 'error');
    render(
      <div>
        <Dropdown menu={{ items }} placement="bottomCenter">
          <button type="button">bottomCenter</button>
        </Dropdown>
        <Dropdown menu={{ items }} placement="topCenter">
          <button type="button">topCenter</button>
        </Dropdown>
      </div>,
    );
    expect(error).toHaveBeenCalledWith(
      expect.stringContaining("[antd: Dropdown] You are using 'bottomCenter'"),
    );
    expect(error).toHaveBeenCalledWith(
      expect.stringContaining("[antd: Dropdown] You are using 'topCenter'"),
    );
    error.mockRestore();
  });

  // zombieJ: when replaced with react test lib, it may be mock fully content
  it('dropdown should support auto adjust placement', () => {
    render(
      <Dropdown menu={{ items }} open>
        <button type="button">button</button>
      </Dropdown>,
    );

    expect(triggerProps.builtinPlacements).toEqual(
      expect.objectContaining({
        bottomLeft: expect.objectContaining({
          overflow: {
            adjustX: 1,
            adjustY: 1,
          },
        }),
      }),
    );
  });

  it('menu item with group', () => {
    jest.useFakeTimers();
    const { container } = render(
      <Dropdown
        trigger={['click']}
        menu={{
          items: [
            {
              label: 'grp',
              type: 'group',
              children: [
                {
                  label: '1',
                  key: 1,
                },
              ],
            },
          ],
        }}
      >
        <a />
      </Dropdown>,
    );

    // Open
    fireEvent.click(container.querySelector('a')!);
    act(() => {
      jest.runAllTimers();
    });

    // Close
    fireEvent.click(container.querySelector('.ant-dropdown-menu-item')!);

    // Force Motion move on
    for (let i = 0; i < 10; i += 1) {
      act(() => {
        jest.runAllTimers();
      });
    }

    // Motion End
    fireEvent.animationEnd(container.querySelector('.ant-slide-up-leave-active')!);

    expect(container.querySelector('.ant-dropdown-hidden')).toBeTruthy();

    jest.useRealTimers();
  });

  it('deprecated warning', () => {
    const errSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

    const { rerender } = render(
      <Dropdown
        visible
        menu={{
          items: [
            {
              label: <div className="bamboo" />,
              key: 'bamboo',
            },
          ],
        }}
      >
        <a />
      </Dropdown>,
    );
    expect(errSpy).toHaveBeenCalledWith(
      'Warning: [antd: Dropdown] `visible` is deprecated which will be removed in next major version, please use `open` instead.',
    );
    rerender(
      <Dropdown
        onVisibleChange={() => {}}
        menu={{
          items: [
            {
              label: <div className="bamboo" />,
              key: 'bamboo',
            },
          ],
        }}
      >
        <a />
      </Dropdown>,
    );
    expect(errSpy).toHaveBeenCalledWith(
      'Warning: [antd: Dropdown] `onVisibleChange` is deprecated which will be removed in next major version, please use `onOpenChange` instead.',
    );
    rerender(
      <Dropdown overlay={<div>menu</div>}>
        <a />
      </Dropdown>,
    );
    expect(errSpy).toHaveBeenCalledWith(
      'Warning: [antd: Dropdown] `overlay` is deprecated. Please use `menu` instead.',
    );

    errSpy.mockRestore();
  });

  it('control dropdown visibility by menu keys when click menu item', () => {
    jest.useFakeTimers();
    const { container } = render(
      <Dropdown
        trigger={['click']}
        closeOnSelectKeys={['close']}
        overlay={
          <Menu
            items={[
              {
                key: '1',
                label: 'menu item 1',
                title: 'menu item 1',
              },
              {
                key: '2',
                label: 'menu item 2',
                title: 'menu item 2',
              },
              {
                key: '3',
                label: 'menu item 3',
                title: 'menu item 3',
              },
              {
                key: 'close',
                danger: true,
                label: 'close',
                title: 'close',
              },
            ]}
          />
        }
      >
        <a />
      </Dropdown>,
    );

    // Open
    fireEvent.click(container.querySelector('a')!);
    act(() => {
      jest.runAllTimers();
    });

    // do not close when click item's key is not include in closeOnSelectKeys
    fireEvent.click(container.querySelector('[title="menu item 1"]')!);

    expect(container.querySelector('.ant-dropdown-hidden')).toBeFalsy();

    // Open
    fireEvent.click(container.querySelector('a')!);
    act(() => {
      jest.runAllTimers();
    });

    // dropdown should be closed when click item's key is include closeOnSelectKeys
    fireEvent.click(container.querySelector('[title="close"]')!);

    // Force Motion move on
    for (let i = 0; i < 10; i += 1) {
      act(() => {
        jest.runAllTimers();
      });
    }

    // Motion End
    fireEvent.animationEnd(container.querySelector('.ant-slide-up-leave-active')!);

    expect(container.querySelector('.ant-dropdown-hidden')).toBeTruthy();

    // =========================================
    // if closeOnSelectKeys pass with empty array,dropdown will not close when click menu item
    const { container: container2 } = render(
      <Dropdown
        trigger={['click']}
        closeOnSelectKeys={[]}
        overlay={
          <Menu
            items={[
              {
                key: '1',
                label: 'menu item 1',
                title: 'menu item 1',
              },
              {
                key: '2',
                label: 'menu item 2',
                title: 'menu item 2',
              },
              {
                key: '3',
                label: 'menu item 3',
                title: 'menu item 3',
              },
              {
                key: 'close',
                danger: true,
                label: 'close',
                title: 'close',
              },
            ]}
          />
        }
      >
        <a />
      </Dropdown>,
    );
    // Open
    fireEvent.click(container2.querySelector('a')!);
    act(() => {
      jest.runAllTimers();
    });

    // do not close when click item's key is not include in closeOnSelectKeys
    fireEvent.click(container2.querySelector('[title="menu item 1"]')!);
    expect(container2.querySelector('.ant-dropdown-hidden')).toBeFalsy();
    fireEvent.click(container2.querySelector('[title="close"]')!);
    expect(container2.querySelector('.ant-dropdown-hidden')).toBeFalsy();
    jest.useRealTimers();
  });
});
