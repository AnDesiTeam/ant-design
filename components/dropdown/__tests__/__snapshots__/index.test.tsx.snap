// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown overlay is function and has custom transitionName 1`] = `
<button
  class="ant-dropdown-trigger ant-dropdown-open"
  type="button"
>
  button
</button>
`;

exports[`Dropdown overlay is string 1`] = `
<button
  class="ant-dropdown-trigger ant-dropdown-open"
  type="button"
>
  button
</button>
`;

exports[`Dropdown rtl render component should be rendered correctly in RTL direction 1`] = `
<span
  class="ant-dropdown-trigger ant-dropdown-rtl"
/>
`;
